app.py

```diff
@@ -31,46 +30,176 @@ def init_app(self):
+        :param entity_id: int for the PublishedFile ID.
+        :return: str path or None depending on whether it successfully resolves the path.
+        """
+
+        # Find the publish so that we can get the path back.
+        published_file = self.sgtk.shotgun.find_one("PublishedFile", [["id", "is", entity_id]], ["code", "path"])
+        # Resolve the path for the local OS
+        try:
+            local_path = sgtk.util.resolve_publish_path(self.sgtk, published_file)
+        except Exception as e:
+            # It might fail to resolve the path to the publish if the published file is an uploaded file
+            # or URL, in which case we revert to the default context based method.
+            self.log_warning("Publish path couldn't be resolved "
+                             "falling back to context based path; reason: %s" % e)
+            local_path = None
+
+        return local_path
+
     def show_in_filesystem(self, entity_type, entity_ids):
         """
         Pop up a filesystem finder window for each folder associated
         with the given entity ids.
         """
         paths = []
-        
+
+        # As the this method relies on the path_cache to find the folders,
+        # we must check we are in sync before continuing.
+        self.sgtk.synchronize_filesystem_structure()
+
         for eid in entity_ids:
-            # Use the path cache to look up all paths linked to the task's entity
-            context = self.tank.context_from_entity(entity_type, eid)
-            paths.extend( context.filesystem_locations )
+
+            file_path_found = False
+
+            if entity_type == "PublishedFile":
+                # If the entity type is a PublishedFile, we should try at first to extract the path of the publish
+                # and open the folder for that. If we can't we will fall back on the context driven path.
+                pub_file_path = self._get_published_file_path(eid)
+                if pub_file_path:
+                    paths.append(pub_file_path)
+                    file_path_found = True
+
+            if not file_path_found:
+                # We've not found path to a specific file for the entity, so we should try to resolve the path
+                # using the path cache instead.
+
+                # Use the path cache to look up all paths linked to the task's entity
+                context = self.sgtk.context_from_entity(entity_type, eid)
+                context_paths = context.filesystem_locations
+
+                if context.step:
+                    # As steps are non project entities and not linked from the step to the other entities
+                    # you won't get step paths back from the context.filesystem_locations.
+                    # If we have a step in the context,
+                    # we should check to see if we can resolve the path more deeply using that.
+                    step_paths = self.sgtk.paths_from_entity("Step", context.step["id"])
```

Comment:
this will potentially return thousands of paths, it doesn't scale.

Could we instead say something like: if we are looking at a step, use shotgun to find all the tasks matching that entity and that step and then get the paths for those tasks. That should result in a constant number of matches even as a production scales up to hundreds of thousands of tasks.

Reply:
Getting the `Task` using the `Step` and entity in the context should be a problem. However, you won't get any paths from the `Task` entities (unless you actually have `Task` folders in your schema). Our default is to just have step folders and not task ones. So I could check for tasks but it would only work if there were task folders as far as I can tell.

Reply:
We need to scale it down so that we don't get all the step paths. Can you get the entity first, and then based on that request the step folders somehow? So that we don't ask for *all* steps, since that doesn't scale.

Reply:
The path cache just doesn't contain anything that can link a step registry to the task it was associated with when the folder was created (understandably as steps can be linked to many tasks.) It also doesn't contain anything that links it to the entity either. I've not looked into how to do this yet, but perhaps I can generate the path for the step against the schema (the same way create folders does) and then maybe even check to see if the generated step folder exists in the path cache.